This following patch was originally found at the debian mingw-w64 team's
mingw-w64 repo located here:
https://salsa.debian.org/mingw-w64-team/mingw-w64.git

Invoke the following in the aforementioned repo to see the original patch:

  $ git show 4974e2c:debian/patches/dlltool-temp-prefix.patch

Description: Specify dlltool's temp prefix
Author: Stephen Kitt <steve@sk2.org>

By default dlltool uses its pid for the object files it generates.
Enforcing its temp prefix allows the files it generates to be
reproducible.

--- a/mingw-w64-crt/Makefile.am
+++ b/mingw-w64-crt/Makefile.am
@@ -36,7 +36,7 @@
   DTDEF32=$(GENLIB) $(DLLTOOLFLAGS32) $(AM_DLLTOOLFLAGS)
   DTDEF64=$(GENLIB) $(DLLTOOLFLAGS64) $(AM_DLLTOOLFLAGS)
 else
-  AM_DLLTOOLFLAGS=-k --as=$(AS) --output-lib $@
+  AM_DLLTOOLFLAGS=-k --as=$(AS) --output-lib $@ --temp-prefix $$(basename $@ .a)
   DLLTOOLFLAGSARM32=-m arm
   DLLTOOLFLAGSARM64=-m arm64
   DLLTOOLFLAGS32=--as-flags=--32 -m i386
